Bottom: 1ca4836dbc6d685024c5d2a5f8289dfecf850d3f
Top:    ea161c78b16e178a00f6817cce3e6c14fc5b5237
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-23 14:41:49 +0100

Fixup gtk hid drawing routines for pour support


---

diff --git a/src/hid/gtk/gtkhid-gl.c b/src/hid/gtk/gtkhid-gl.c
index f0bcbf5..9159eed 100644
--- a/src/hid/gtk/gtkhid-gl.c
+++ b/src/hid/gtk/gtkhid-gl.c
@@ -1202,6 +1202,10 @@ DrawPlainPolygon (LayerTypePtr Layer, PolygonTypePtr Polygon, const BoxType *dra
   if (!Polygon->Clipped)
     return;
 
+  /* Re-use HOLEFLAG to cut out islands */
+  if (TEST_FLAG (HOLEFLAG, Polygon))
+    return;
+
   if (TEST_FLAG (SELECTEDFLAG, Polygon))
     color = Layer->SelectedColor;
   else if (TEST_FLAG (FOUNDFLAG, Polygon))
@@ -1264,6 +1268,38 @@ poly_callback_clearing (const BoxType * b, void *cl)
 }
 
 static int
+pour_callback_no_clear (const BoxType * b, void *cl)
+{
+  struct poly_info *i = (struct poly_info *) cl;
+  PourType *pour = (PourType *)b;
+
+  DrawPour (i->Layer, pour);
+
+  if (pour->PolygonN)
+    {
+      r_search (pour->polygon_tree, i->drawn_area, NULL, poly_callback_no_clear, i);
+    }
+
+  return 1;
+}
+
+  static int
+pour_callback_clearing (const BoxType * b, void *cl)
+{
+  struct poly_info *i = (struct poly_info *) cl;
+  PourType *pour = (PourType *)b;
+
+  DrawPour (i->Layer, pour);
+
+  if (pour->PolygonN)
+    {
+      r_search (pour->polygon_tree, i->drawn_area, NULL, poly_callback_clearing, i);
+    }
+
+  return 1;
+}
+
+static int
 clearPin_callback (const BoxType * b, void *cl)
 {
   PinType *pin = (PinTypePtr) b;
@@ -1393,11 +1429,11 @@ DrawLayerGroup (int group, const BoxType * screen)
       }
 
       /* draw all polygons on this layer */
-      if (Layer->PolygonN) {
+      if (Layer->PourN) {
         info.Layer = Layer;
         info.drawn_area = screen;
-        r_search (Layer->polygon_tree, screen, NULL, poly_callback_no_clear, &info);
-        r_search (Layer->polygon_tree, screen, NULL, poly_callback_clearing, &info);
+        r_search (Layer->pour_tree, screen, NULL, pour_callback_no_clear, &info);
+        r_search (Layer->pour_tree, screen, NULL, pour_callback_clearing, &info);
 
         /* HACK: Subcomposite polygons separately from other layer primitives */
         /* Reset the compositing */
